src/centralized_telescope.rs

```diff
@@ -0,0 +1,419 @@
+            Setup {
+                n_p: params.n_p,
+                u,
+                r: (lambda_rel / lambda_rel2).ceil() as u64,
+                d: d as u64,
+                q: 2.0 * (lambda_rel2 + 2.0) / (d * loge),
+                b: (((lambda_rel2 + 2.0 + u_f64.log2()) / (lambda_rel2 + 2.0))
+                    * (3.0 * u_f64 * d / 4.0)
+                    + d
+                    + u_f64)
+                    .floor() as u64,
+            }
+        } else {
+            // Case 2, Theorem 13, ie λ^3 > n_p > λ^2
+            let lambda_rel1 = lambda_rel.min(s1);
+            let lbar = (lambda_rel1 + 7.0) / loge;
+            let d = (16.0 * u_f64 * lbar).ceil();
+            debug_assert!(n_p_f64 >= d * d / (9.0 * lbar));
+
+            let w = compute_w(u_f64, lambda_rel1);
+            Setup {
+                n_p: params.n_p,
+                u,
+                r: (lambda_rel / lambda_rel1).ceil() as u64,
+                d: d as u64,
+                q: 2.0 * lbar / d,
+                b: ((w * lbar / d + 1.0)
+                    * (2.0 * u_f64 * w * lbar / n_p_f64 + 7.0 * u_f64 / w).exp()
+                    * d
+                    * u_f64
+                    + d)
+                    .floor() as u64,
+            }
+        }
+    }
+}
+
+/// Round parameters
+#[derive(Debug, Clone)]
+pub struct Round {
+    /// Proof counter
+    v: u64,
+    /// Proof 2nd counter
+    t: u64,
+    // Round candidate tuple
+    s_list: Vec<Element>,
+    /// Round candidate hash
+    h: Hash,
+    /// Round candidate hash mapped to [1, n_p]
+    h_u64: u64,
+    /// Approximate size of set Sp to lower bound
+    n_p: u64,
+}
+
+impl Round {
+    /// Oracle producing a uniformly random value in [1, n_p] used for round candidates
+    /// We also return hash(data) to follow the optimization presented in Section 3.3
+    fn h1(input: &[Vec<u8>], n_p: u64) -> (Hash, u64) {
+        let mut data = vec!["Telescope-H1".as_bytes().to_vec()];
+        for i in input {
+            data.push(i.to_vec());
+        }
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(&data);
+        (digest, utils::sample_uniform(&digest, n_p))
+    }
+
+    /// Output a round from a proof counter and n_p
+    /// Initilialises the hash with H1(t) and random value as oracle(H1(t), n_p)
+    pub fn new(v: u64, t: u64, n_p: u64) -> Round {
+        let mut data = vec![v.to_ne_bytes().to_vec()];
+        data.push(t.to_ne_bytes().to_vec());
+        let (h, h_u64) = Round::h1(&data, n_p);
+        Round {
+            v,
+            t,
+            s_list: vec![],
+            h,
+            h_u64,
+            n_p,
+        }
+    }
+
+    /// Updates a round with an element of S_p
+    /// Replaces the hash $h$ with $h' = H1(h, s)$ and the random value as oracle(h', n_p)
+    pub fn update(r: &Round, s: Element) -> Round {
+        let mut s_list = r.s_list.clone();
+        s_list.push(s);
+        let mut data = vec![r.h.clone().to_vec()];
```

Comment:
as before, if you want to avoid creating this new data and cloning stuff, you can use the `Blake2b::update()`. That way, you can directly use the data available in `r` to update the hasher (same idea as the comment in line 196.